File: code/backend/java/adapter-rest/src/main/java/games/paths/adapters/rest/controller/match/MatchController.java
package games.paths.adapters.rest.controller.match;

import games.paths.adapters.rest.dto.MatchCreateRequest;
import games.paths.adapters.rest.dto.MatchInfoResponse;
import games.paths.adapters.rest.dto.MatchSummaryResponse;
import games.paths.core.model.match.MatchCreateCommand;
import games.paths.core.model.match.MatchDetail;
import games.paths.core.model.match.MatchSummary;
import games.paths.core.port.match.MatchCommandPort;
import games.paths.core.port.match.MatchQueryPort;
import games.paths.core.service.security.CsrfTokenService;
import games.paths.core.service.security.RateLimitService;

import jakarta.servlet.http.HttpServletRequest;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * MatchController - REST adapter for single-player match operations.
 *
 * <ul>
 *   <li>POST /api/matches              — create a new match</li>
 *   <li>GET  /api/matches              — list current user matches</li>
 *   <li>GET  /api/match/&#123;uuid&#125;/info  — match details (state + registry)</li>
 * </ul>
 *
 * <p>Step 19 — see {@code documentation_v0/Step19_SinglePlayerMatchCreation.md}.</p>
 */
@RestController
public class MatchController {

    private final MatchCommandPort matchCommandPort;
    private final MatchQueryPort matchQueryPort;

    static final String RATE_BUCKET = "match";

    private final RateLimitService rateLimitService;
    private final int matchPerIp;
    private final CsrfTokenService csrfTokenService;

    public MatchController(MatchCommandPort matchCommandPort, MatchQueryPort matchQueryPort) {
        this(matchCommandPort, matchQueryPort, null, 0, null);
    }

    /** v0.37.7 — Step 41: the match bucket of the rate limiter and the CSRF check on creation. */
    @org.springframework.beans.factory.annotation.Autowired
    public MatchController(MatchCommandPort matchCommandPort, MatchQueryPort matchQueryPort,
                           RateLimitService rateLimitService,
                           @org.springframework.beans.factory.annotation.Value("${game.security.rate-limit.match-per-ip:0}") int matchPerIp,
                           CsrfTokenService csrfTokenService) {
        this.matchCommandPort = matchCommandPort;
        this.matchQueryPort = matchQueryPort;
        this.rateLimitService = rateLimitService;
        this.matchPerIp = matchPerIp;
        this.csrfTokenService = csrfTokenService;
    }

    @PostMapping("/api/matches")
    public ResponseEntity<Object> createMatch(@RequestBody(required = false) MatchCreateRequest body,
                                              HttpServletRequest request) {
        String userUuid = (String) request.getAttribute("userUuid");
        if (userUuid == null || userUuid.isBlank()) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED",
                    "User identity is missing from the request");
        }
        // Step 41 — the CSRF token issued with this access token must come back as a header
        if (csrfTokenService != null && csrfTokenService.isEnforced()) {
            String presented = request.getHeader(CsrfTokenService.HEADER);
            if (presented == null || presented.isBlank()) {
                return error(HttpStatus.FORBIDDEN, "CSRF_TOKEN_MISSING",
                        CsrfTokenService.HEADER + " header is required to create a match");
            }
            if (!csrfTokenService.matches(bearerOf(request), presented)) {
                return error(HttpStatus.FORBIDDEN, "CSRF_TOKEN_INVALID",
                        CsrfTokenService.HEADER + " does not match the access token");
            }
        }
        // Step 41 — at most match-per-ip new matches per source address and window
        if (rateLimitService != null && matchPerIp > 0) {
            String ip = RateLimitService.clientIp(request.getHeader("X-Forwarded-For"),
                    request.getRemoteAddr());
            RateLimitService.Verdict verdict = rateLimitService.tryAcquire(RATE_BUCKET, ip, matchPerIp);
            if (!verdict.allowed()) {
                return ResponseEntity.status(HttpStatus.TOO_MANY_REQUESTS)
                        .header("Retry-After", String.valueOf(verdict.retryAfterSeconds()))
                        .body(errorBody("RATE_LIMITED", "Too many matches created from this address, retry in "
                                + verdict.retryAfterSeconds() + " seconds", verdict.retryAfterSeconds()));
            }
        }
        if (body == null || isBlank(body.getStoryUuid()) || isBlank(body.getDifficultyUuid())) {
            return error(HttpStatus.BAD_REQUEST, "INVALID_INPUT",
                    "storyUuid and difficultyUuid are required");
        }

        MatchCreateCommand command = new MatchCreateCommand(
                userUuid,
                body.getStoryUuid(),
                body.getDifficultyUuid(),
                body.getName(),
                body.getCharacterTemplateUuid(),
                body.getClassUuid(),
                body.getTraitUuids(),
                body.getSinglePlayer(),
                body.getTurnstileToken(),
                request.getRemoteAddr(),
                body.getRngSeed());

        try {
            MatchSummary created = matchCommandPort.createMatch(command);
            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(MatchSummaryResponse.fromModel(created));
        } catch (MatchCommandPort.MatchCreationException ex) {
            return error(mapStatus(ex.getCode()), ex.getCode().name(), ex.getMessage());
        }
    }

    @GetMapping("/api/matches")
    public ResponseEntity<Object> listMatches(HttpServletRequest request) {
        String userUuid = (String) request.getAttribute("userUuid");
        if (userUuid == null || userUuid.isBlank()) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED",
                    "User identity is missing from the request");
        }
        List<MatchSummary> models = matchQueryPort.listUserMatches(userUuid);
        List<MatchSummaryResponse> result = models.stream()
                .map(MatchSummaryResponse::fromModel)
                .collect(Collectors.toList());
        return ResponseEntity.ok(result);
    }

    @GetMapping("/api/match/{uuidMatch}/info")
    public ResponseEntity<Object> getMatchInfo(@PathVariable String uuidMatch,
                                               @RequestParam(value = "lang", defaultValue = "en") String lang,
                                               HttpServletRequest request) {
        String userUuid = (String) request.getAttribute("userUuid");
        if (userUuid == null || userUuid.isBlank()) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED",
                    "User identity is missing from the request");
        }
        if (isBlank(uuidMatch)) {
            return error(HttpStatus.BAD_REQUEST, "INVALID_INPUT",
                    "Match uuid is required");
        }
        MatchDetail detail = matchQueryPort.getMatchInfo(uuidMatch, userUuid, lang);
        if (detail == null) {
            return error(HttpStatus.NOT_FOUND, "MATCH_NOT_FOUND",
                    "Match not found or not accessible");
        }
        return ResponseEntity.ok(MatchInfoResponse.fromModel(detail));
    }

    /**
     * PATCH /api/match/{uuidMatch}/end/{uuidEvent} — Step 20.1.
     * Completes a match (sets status to ENDED) when the supplied event uuid is
     * the story's end-game event. Returns 406 Not Acceptable when the event is
     * not the configured end-game trigger. The idEventEndGame value is never
     * exposed in any response payload.
     */
    @PatchMapping("/api/match/{uuidMatch}/end/{uuidEvent}")
    public ResponseEntity<Object> endMatch(@PathVariable String uuidMatch,
                                           @PathVariable String uuidEvent,
                                           HttpServletRequest request) {
        String userUuid = (String) request.getAttribute("userUuid");
        if (userUuid == null || userUuid.isBlank()) {
            return error(HttpStatus.UNAUTHORIZED, "UNAUTHENTICATED",
                    "User identity is missing from the request");
        }
        if (isBlank(uuidMatch) || isBlank(uuidEvent)) {
            return error(HttpStatus.BAD_REQUEST, "INVALID_INPUT",
                    "Match uuid and event uuid are required");
        }
        MatchCommandPort.EndMatchOutcome outcome =
                matchCommandPort.endMatch(uuidMatch, uuidEvent, userUuid);
        switch (outcome) {
            case COMPLETED:
                Map<String, Object> body = new LinkedHashMap<>();
                body.put("status", "ENDED");
                body.put("uuid", uuidMatch);
                return ResponseEntity.ok(body);
            case NOT_ACCEPTABLE:
                return error(HttpStatus.NOT_ACCEPTABLE, "EVENT_NOT_END_GAME",
                        "The supplied event is not the end-game event for this match");
            case NOT_FOUND:
            default:
                return error(HttpStatus.NOT_FOUND, "MATCH_NOT_FOUND",
                        "Match not found or not accessible");
        }
    }

    private static boolean isBlank(String s) {
        return s == null || s.isBlank();
    }

    private static ResponseEntity<Object> error(HttpStatus status, String code, String message) {
        return ResponseEntity.status(status).body(errorBody(code, message, null));
    }

    private static Map<String, Object> errorBody(String code, String message, Long retryAfterSeconds) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("error", code);
        body.put("message", message);
        if (retryAfterSeconds != null) {
            body.put("retryAfterSeconds", retryAfterSeconds);
        }
        body.put("timestamp", System.currentTimeMillis());
        return body;
    }

    /** The raw bearer the filter already validated — the CSRF token is bound to it. */
    private static String bearerOf(HttpServletRequest request) {
        String header = request.getHeader("Authorization");
        return header != null && header.startsWith("Bearer ") ? header.substring(7).trim() : null;
    }

    private static HttpStatus mapStatus(MatchCommandPort.MatchCreationException.Code code) {
        return switch (code) {
            case STORY_NOT_FOUND, DIFFICULTY_NOT_FOUND, USER_NOT_FOUND -> HttpStatus.NOT_FOUND;
            case USER_BANNED -> HttpStatus.FORBIDDEN;
            case MAINTENANCE_MODE -> HttpStatus.SERVICE_UNAVAILABLE;
            case ACTIVE_MATCH_ALREADY_EXISTS -> HttpStatus.CONFLICT;
            case STORY_HAS_NO_LOCATIONS, INVALID_INPUT, TURNSTILE_VALIDATION_FAILED,
                 TRAIT_NOT_FOUND, TRAIT_DUPLICATED,
                 TRAIT_NOT_COMPATIBLE, TRAIT_COST_EXCEEDED,
                 TRAIT_NOT_SELECTABLE -> HttpStatus.BAD_REQUEST;
        };
    }
}

File: code/backend/php/src/Adapter/Rest/Matches/MatchController.php
<?php

namespace Games\Paths\Adapter\Rest\Matches;

use Games\Paths\Core\Domain\Matches\MatchCreateCommand;
use Games\Paths\Core\Domain\Matches\MatchCreationException;
use Games\Paths\Core\Domain\Matches\MatchStatuses;
use Games\Paths\Core\Port\Matches\MatchCommandPort;
use Games\Paths\Core\Port\Matches\MatchQueryPort;
use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;

class MatchController
{
    private const STATUS_BY_CODE = [
        MatchCreationException::INVALID_INPUT => 400,
        MatchCreationException::STORY_HAS_NO_LOCATIONS => 400,
        MatchCreationException::TURNSTILE_VALIDATION_FAILED => 400,
        MatchCreationException::STORY_NOT_FOUND => 404,
        MatchCreationException::DIFFICULTY_NOT_FOUND => 404,
        MatchCreationException::USER_NOT_FOUND => 404,
        MatchCreationException::USER_BANNED => 403,
        MatchCreationException::MAINTENANCE_MODE => 503,
    ];

    public function __construct(
        private readonly MatchCommandPort $commandPort,
        private readonly MatchQueryPort $queryPort
    ) {
    }

    public function createMatch(Request $request, Response $response): Response
    {
        $userUuid = (string)($request->getAttribute('userUuid') ?? '');
        if ($userUuid === '') {
            return $this->error($response, 'UNAUTHENTICATED', 'User identity is missing', 401);
        }
        $body = (array)($request->getParsedBody() ?? []);
        $storyUuid = (string)($body['storyUuid'] ?? '');
        $difficultyUuid = (string)($body['difficultyUuid'] ?? '');
        if ($storyUuid === '' || $difficultyUuid === '') {
            return $this->error($response, 'INVALID_INPUT', 'storyUuid and difficultyUuid are required', 400);
        }

        $traitUuids = $body['traitUuids'] ?? [];
        $serverParams = $request->getServerParams();
        $remoteIp = $serverParams['REMOTE_ADDR'] ?? null;
        $command = new MatchCreateCommand(
            userUuid: $userUuid,
            storyUuid: $storyUuid,
            difficultyUuid: $difficultyUuid,
            name: $body['name'] ?? null,
            characterTemplateUuid: $body['characterTemplateUuid'] ?? null,
            classUuid: $body['classUuid'] ?? null,
            traitUuids: is_array($traitUuids) ? $traitUuids : [],
            singlePlayer: isset($body['singlePlayer']) ? (int)$body['singlePlayer'] : null,
            turnstileToken: $body['turnstileToken'] ?? null,
            remoteIp: $remoteIp
        );

        try {
            $summary = $this->commandPort->createMatch($command);
        } catch (MatchCreationException $e) {
            $status = self::STATUS_BY_CODE[$e->getCodeId()] ?? 400;
            return $this->error($response, $e->getCodeId(), $e->getMessage(), $status);
        }

        $response->getBody()->write(json_encode($summary->toArray()));
        return $response->withStatus(201)->withHeader('Content-Type', 'application/json');
    }

    public function listMatches(Request $request, Response $response): Response
    {
        $userUuid = (string)($request->getAttribute('userUuid') ?? '');
        if ($userUuid === '') {
            return $this->error($response, 'UNAUTHENTICATED', 'User identity is missing', 401);
        }
        $matches = $this->queryPort->listUserMatches($userUuid);
        $body = array_map(fn($m) => $m->toArray(), $matches);
        $response->getBody()->write(json_encode($body));
        return $response->withStatus(200)->withHeader('Content-Type', 'application/json');
    }

    public function listAllMatches(Request $request, Response $response): Response
    {
        // The admin role is enforced by JwtAuthenticationMiddleware for /api/admin/ paths.
        $matches = $this->queryPort->listAllMatches();
        $body = array_map(fn($m) => $m->toArray(), $matches);
        $response->getBody()->write(json_encode($body));
        return $response->withStatus(200)->withHeader('Content-Type', 'application/json');
    }

    public function getMatchInfo(Request $request, Response $response, array $args): Response
    {
        $userUuid = (string)($request->getAttribute('userUuid') ?? '');
        if ($userUuid === '') {
            return $this->error($response, 'UNAUTHENTICATED', 'User identity is missing', 401);
        }
        $matchUuid = (string)($args['uuidMatch'] ?? '');
        if ($matchUuid === '') {
            return $this->error($response, 'INVALID_INPUT', 'Match uuid is required', 400);
        }
        $detail = $this->queryPort->getMatchInfo($matchUuid, $userUuid);
        if ($detail === null) {
            return $this->error($response, 'MATCH_NOT_FOUND', 'Match not found or not accessible', 404);
        }
        $response->getBody()->write(json_encode($detail->toArray()));
        return $response->withStatus(200)->withHeader('Content-Type', 'application/json');
    }

    /**
     * GET /api/admin/matches/statuses — valid match statuses, each flagged
     * "terminal" when a match in that status is stopped (deletable).
     */
    public function listMatchStatuses(Request $request, Response $response): Response
    {
        $body = array_map(
            static fn(string $s): array => ['value' => $s, 'terminal' => MatchStatuses::isTerminal($s)],
            MatchStatuses::ALL
        );
        return $this->ok($response, $body);
    }

    /**
     * GET /api/admin/matches/{uuidMatch}/info — full match detail for the
     * admin console, without the per-user ownership check enforced by
     * GET /api/match/{uuidMatch}/info.
     */
    public function getAdminMatchInfo(Request $request, Response $response, array $args): Response
    {
        $matchUuid = (string)($args['uuidMatch'] ?? '');
        if ($matchUuid === '') {
            return $this->error($response, 'INVALID_INPUT', 'Match uuid is required', 400);
        }
        $detail = $this->queryPort->getMatchInfoForAdmin($matchUuid);
        if ($detail === null) {
            return $this->error($response, 'MATCH_NOT_FOUND', 'Match not found: ' . $matchUuid, 404);
        }
        $response->getBody()->write(json_encode($detail->toArray()));
        return $response->withStatus(200)->withHeader('Content-Type', 'application/json');
    }

    /** PUT /api/admin/matches/{uuidMatch} — admin update of status and/or name. */
    public function updateMatch(Request $request, Response $response, array $args): Response
    {
        $uuid = (string)($args['uuidMatch'] ?? '');
        $body = (array)($request->getParsedBody() ?? []);
        $status = array_key_exists('status', $body) ? $body['status'] : null;
        $name = array_key_exists('name', $body) ? $body['name'] : null;
        if ($status === null && $name === null) {
            return $this->error($response, 'INVALID_INPUT',
                'At least one of status or name must be provided', 400);
        }
        return $this->applyUpdate($response, $uuid, $status, $name);
    }

    /** POST /api/admin/matches/{uuidMatch}/stop — sets the match status to ENDED. */
    public function stopMatch(Request $request, Response $response, array $args): Response
    {
        return $this->applyUpdate($response, (string)($args['uuidMatch'] ?? ''), MatchStatuses::ENDED, null);
    }

    /** POST /api/admin/matches/{uuidMatch}/pause — sets the match status to PAUSED. */
    public function pauseMatch(Request $request, Response $response, array $args): Response
    {
        return $this->applyUpdate($response, (string)($args['uuidMatch'] ?? ''), MatchStatuses::PAUSED, null);
    }

    /** POST /api/admin/matches/{uuidMatch}/resume — sets the match status to RUNNING. */
    public function resumeMatch(Request $request, Response $response, array $args): Response
    {
        return $this->applyUpdate($response, (string)($args['uuidMatch'] ?? ''), MatchStatuses::RUNNING, null);
    }

    /**
     * DELETE /api/admin/matches/{uuidMatch} — deletes a match. Only matches in
     * a terminal status (ENDED / GAMEOVER) may be deleted.
     */
    public function deleteMatch(Request $request, Response $response, array $args): Response
    {
        $uuid = (string)($args['uuidMatch'] ?? '');
        $outcome = $this->commandPort->deleteMatch($uuid);
        if ($outcome === 'DELETED') {
            return $this->ok($response, ['status' => 'DELETED', 'uuid' => $uuid]);
        }
        if ($outcome === 'NOT_STOPPED') {
            return $this->error($response, 'MATCH_NOT_STOPPED',
                'Only stopped matches (ENDED or GAMEOVER) can be deleted', 409);
        }
        return $this->error($response, 'MATCH_NOT_FOUND', 'Match not found: ' . $uuid, 404);
    }

    private function applyUpdate(Response $response, string $uuid, ?string $status, ?string $name): Response
    {
        $outcome = $this->commandPort->updateMatch($uuid, $status, $name);
        if ($outcome === 'UPDATED') {
            return $this->ok($response, ['status' => 'UPDATED', 'uuid' => $uuid]);
        }
        if ($outcome === 'INVALID_STATUS') {
            return $this->error($response, 'INVALID_STATUS',
                'status must be one of ' . implode(', ', MatchStatuses::ALL), 400);
        }
        return $this->error($response, 'MATCH_NOT_FOUND', 'Match not found: ' . $uuid, 404);
    }

    private function ok(Response $response, array $data): Response
    {
        $response->getBody()->write(json_encode($data));
        return $response->withStatus(200)->withHeader('Content-Type', 'application/json');
    }

    private function error(Response $response, string $code, string $message, int $status): Response
    {
        $body = json_encode([
            'error' => $code,
            'message' => $message,
            'timestamp' => (int)(microtime(true) * 1000),
        ]);
        $response->getBody()->write($body);
        return $response->withStatus($status)->withHeader('Content-Type', 'application/json');
    }
}
